pkg/scheduler/framework/cycle_state_test.go

```diff
@@ -31,44 +34,98 @@ func (f *fakeData) Clone() StateData {
+// The given data is used in stored fakeData.
+func createCycleStateWithFakeData(data string, recordPluginMetrics bool) *CycleState {
+	c := NewCycleState()
+	c.Write(key, &fakeData{
+		data: data,
+	})
+	c.SetRecordPluginMetrics(recordPluginMetrics)
+	return c
+}
 
-	valueCopy, err := stateCopy.Read(key)
-	if err != nil {
-		t.Errorf("failed to read copied value: %v", err)
-	}
-	if v, ok := valueCopy.(*fakeData); ok && v.data != data1 {
-		t.Errorf("clone failed, got %q, expected %q", v.data, data1)
+func TestCycleStateClone(t *testing.T) {
+	tests := []struct {
+		name            string
+		state           *CycleState
+		wantClonedState *CycleState // clonedState is CycleState cloned by Clone().
+	}{
+		{
+			name:            "clone with recordPluginMetrics true",
+			state:           createCycleStateWithFakeData("data", true),
+			wantClonedState: createCycleStateWithFakeData("data", true),
+		},
+		{
+			name:            "clone with recordPluginMetrics false",
+			state:           createCycleStateWithFakeData("data", false),
+			wantClonedState: createCycleStateWithFakeData("data", false),
+		},
+		{
+			name:            "clone with nil CycleState",
+			state:           nil,
+			wantClonedState: nil,
+		},
 	}
 
-	originalValue.data = data2
-	original, err := state.Read(key)
-	if err != nil {
-		t.Errorf("failed to read original value: %v", err)
-	}
-	if v, ok := original.(*fakeData); ok && v.data != data2 {
-		t.Errorf("original value should change, got %q, expected %q", v.data, data2)
-	}
+	cmpOpts := []cmp.Option{
+		cmp.AllowUnexported(CycleState{}),
+		cmp.Comparer(func(a, b *sync.Map) bool {
+			isEqual := true
+
+			countA := 0
+			a.Range(func(k, v1 interface{}) bool {
+				countA++
+				v2, ok := b.Load(k)
+				if !ok {
+					isEqual = false
+					return false
+				}
+
+				if !cmp.Equal(v1, v2, cmp.AllowUnexported(fakeData{})) {
+					isEqual = false
+					return false
+				}
+				return true
+			})
 
-	if v, ok := valueCopy.(*fakeData); ok && v.data != data1 {
-		t.Errorf("cloned copy should not change, got %q, expected %q", v.data, data1)
+			countB := 0
+			b.Range(func(k, _ interface{}) bool {
+				countB++
+				return true
+			})
+
+			if countA != countB {
+				return false
+			}
+			return isEqual || countA != countB
+		}),
 	}
-}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			state := tt.state
+			stateCopy := state.Clone()
+			if state == nil {
+				// in this case, tt.wantClonedState should be nil and we can compare without cmp.Equal.
+				if stateCopy != tt.wantClonedState {
+					t.Errorf("Unexpected cloned state: want %v, but got %v", tt.wantClonedState, stateCopy)
+				}
+				// when state is nil, we don't need to run the rest check.
+				return
+			}
+
+			// we need to compare *sync.Map,
+			// because we cannot define cmp.Comparer for sync.Map or CycleState due to the issue from go vet.
+			if diff := cmp.Diff(&state.storage, &stateCopy.storage, cmpOpts...); diff != "" {
```

Comment:
Now I think nothing is checking `recordPluginMetrics`?

is it a lot of changes if you make CycleState.storage a pointer?

Reply:
Fixed.